File: FindFirstAndLastElement.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class FindFirstAndLastElement {
    public int[] searchRange(int[] nums, int target) {
        if(nums == null || nums.length == 0) return new int[]{-1, -1};
        if (target > nums[nums.length - 1] || target < nums[0]) return new int[]{-1, -1};
        
        int low = 0;
        int high = nums.length - 1;
        
        int first = binarySearchFirst(nums, low, high, target);
        if(first == -1) return new int[]{-1, -1};
        int last = binarySearchLast(nums, low, high, target);
        return new int[]{ first, last };
    }
    
    public int binarySearchFirst(int[] nums, int low, int high, int target){
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == 0 || nums[mid - 1] < nums[mid]){
                    return mid;
                }
                else{
                    high = mid - 1;
                }
            }
            else if(nums[mid]>target){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1;
    }
    
    public int binarySearchLast(int[] nums, int low, int high, int target){
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == nums.length-1 || nums[mid + 1] > nums[mid]){
                    return mid;
                }
                else{
                    low = mid + 1;
                }
            }
            else if(nums[mid]>target){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: find_first_and_last_element.py
# avg TimeComplexity O(log n )
# SpaceComplexity O(1)

class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        first = -1
        def binarysearchFirst(nums, target):
            low = 0
            high = len(nums)-1
            while low <= high:
                mid = (low + high)//2
                if nums[mid] == target:
                    if mid == 0 or nums[mid-1] != nums[mid]:
                        return mid
                    else:
                        high = mid-1
                elif nums[mid] > target:
                    high = mid-1
                else:
                    low = mid+1
            return -1
        
        def binarysearchLast(nums, target):
            low = first
            high = len(nums)-1
            while low <= high:
                mid = (low + high)//2
                if nums[mid] == target:
                    if mid == len(nums)-1 or nums[mid+1] != nums[mid]:
                        return mid
                    else:
                        low = mid + 1
                elif nums[mid] > target:
                    high = mid-1
                else:
                    low = mid+1
            return -1


        if len(nums) == 0: return [-1,-1]
        if nums[0] > target or nums[-1] < target: return [-1,-1]
        first = binarysearchFirst(nums, target)
        if first == -1: return [-1,-1]
        last = binarysearchLast(nums, target)
        return [first, last]